File: code/1493.longest-subarray-of-1-s-after-deleting-one-element.php
<?php
/*
 * @lc app=leetcode id=1493 lang=php
 *
 * [1493] Longest Subarray of 1's After Deleting One Element
 */

// @lc code=start
class Solution {

    /**
     * @param Integer[] $nums
     * @return Integer
     */
    function longestSubarray($nums) {
        $max = 0;
        $useZero = false;
        $count = 0;
        $length = count($nums);
        for ($i = 0; $i < $length; $i++) {
            if ($nums[$i] === 0) {
                if ($useZero) {
                    while (true) {
                        if ($nums[$i - $count--] === 0) {
                            break;
                        }
                    }
                } else {
                    $useZero = true;
                }
            }
            $count++;

            $max = $count > $max ? $count : $max;
        }

        // 0は除外する
        if ($useZero) $max--;

        return $length === $max ? $max - 1 : $max;
    }
}
// @lc code=end

File: code/1493.longest-subarray-of-1-s-after-deleting-one-element.go
package code

/*
 * @lc app=leetcode id=1493 lang=golang
 *
 * [1493] Longest Subarray of 1's After Deleting One Element
 */

// @lc code=start
func longestSubarray(nums []int) int {
	maxWindow := 0
	curSum := 0
	start := 0
	for end := 0; end < len(nums); end++ {
		curSum += nums[end]

		if end-start+1-curSum <= 1 {
			maxWindow = max(maxWindow, end-start)
		} else {
			curSum -= nums[start]
			start++
		}
	}

	return maxWindow
}

// @lc code=end
